### DISCUSSION TITLE
How to pass additional # parameters to AR QuickLook when auto generating USDZ blob?

### DISCUSSION THREAD

**Discussion by milesgreen** [google/model-viewer#3425] [Ideas]
With upcoming iOS fixes (https://github.com/google/model-viewer/issues/3300) it's going to be possible to properly pass # parameters to AR QuickLook when auto generating USDZ blob. 

Indeed, in Safari iOS 15.5 beta 4, when setting ModelViewer to fixed scaling, this now works, as the #allowsContentScaling=0 parameter is appended to the blob url and works correctly. 

So, how might we be able to pass in other QuickLook parameters when auto generating USDZ?

When we have a full USDZ file path, we can append any number of QuickLook parameters to the URL in `ios-src`.

But when generating USDZ, there is no file path or `ios-src`, and so no place to provide these additional parameters. 

Understanding not wanting to pollute ModelViewer with QuickLook-specific parameters.

Maybe one solution would be to allow `ios-src` to be set as just a url # fragment of one or more QuickLook parameters?

Or is there another / better way?

**Comment by elalish** [comment]
Good call. Currently the absence of `ios-src` triggers auto-generation, so that might get confusing. How about appending the parameters to `src`? I believe that's how we can set special SV options currently. Are there any name collisions currently between the SV and QL APIs that don't have matching behavior?

**Reply by milesgreen** [reply]
Interesting. That might work. 

QuickLook parameters include:

**allowsContentScaling**
`0` prevents scaling.

**applePayButtonType**
Predefined Apple Pay button to display in banner. Choice of: `plain / pay / buy / check-out / book / donate / subscribe`

**callToAction**
Custom button label to display in banner, rather than an Apple Pay button

**canonicalWebPageURL**
Specify a URL to use in the share widget from within AR

**checkoutTitle**
String to display as the banner title

**checkoutSubtitle**
String. Line of copy to display under the Title in the banner.

**custom**
URL to HTML page to display in the banner space (rather than using the above parameters).

**customHeight**
String. When displaying custom banner, customHeight can be set to `small`, `medium`, `large` to set the banner height.

**price**
String. Price to be displayed in banner.

I'm not sure of all the SceneViewer parameters but can check. 

Also, I _think_ that QuickLook will simply ignore any invalid parameters. But need to check that.

**Reply by milesgreen** [reply]
SceneViewer parameters include:

enable_vertical_placement
file
link
mode
resizable
S.browser_fallback_url
sound
title

No apparent collisions.

**Reply by elalish** [reply]
Care to try a PR?

**Reply by milesgreen** [reply]
Erm... I wish I could. I don't know I have the chops for that... It'd be a learning curve. 

The main point to work on this would be here, right:

https://github.com/google/model-viewer/blob/0ec49c52a208b0f2bd258660616fcef3bc1fc89c/packages/model-viewer/src/features/ar.ts#L368

**Reply by elalish** [reply]
Correct - and learning is fun, right? 😁 Feel free to put up a draft PR if you need help with anything. The advantage of you taking point on this is that it'll get tested properly, and will also get done sooner (since I'm a bit underwater at the moment). Thanks!

**Reply by milesgreen** [reply]
Heh. Very true. Ok. Let me try to take a swing at it.

**Comment by rawwerks** [comment]
i'm curious: has anyone used this approach to reproduce this kind of AR "buy button" with modelviewer's auto-generated USDZ?

![image](https://user-images.githubusercontent.com/19483938/204055909-f5e21fb2-ea40-4331-80b6-3856b1c4a8a0.png)

from https://developer.apple.com/documentation/arkit/adding_an_apple_pay_button_or_a_custom_action_in_ar_quick_look

(i'm wondering if this approach is a solution to https://github.com/google/model-viewer/discussions/2708)

**Reply by milesgreen** [reply]
Yup. This is the solution you're looking for ;)

**Comment by milesgreen** [comment]
For future reference, this was addressed with https://github.com/google/model-viewer/pull/3434

**Comment by Sov3rain** [comment]
Hey everyone! Re-opening the discussion here, because as of today, it seems like opening an AR Quick Look session with an auto generated USDZ file from a blob is completely ignoring the hash, hence ignoring the parameters that are necessary to display a custom action button.

You can try to open the third example (_Mutliple QuickLook params_) from this page: https://modelviewer-auto-usdz-params.glitch.me/

You should normally see a custom button in place of the "take picture" button.
Is this something that Apple is responsible of again?

**Reply by PixelPartner** [reply]
I have the impression that just the ApplePay related hashes have been deprecated (by Apple last year).

**Reply by Sov3rain** [reply]
Do you have a working example of using AR QuickLook with custum action button? On my side I can't get it to work really.

**Comment by PixelPartner** [comment]
Sorry, noVon meinem iPhone gesendetAm 27.02.2024 um 17:30 schrieb Sov3rain ***@***.***>:﻿
Do you have a working example of using AR QuickLook with custum action button? On my side I can't get it to work really.

**Comment by PixelPartner** [comment]
My post is based on experiences with the # parameters unrelated to to modelViewerVon meinem iPhone gesendetAm 27.02.2024 um 17:30 schrieb Sov3rain ***@***.***>:﻿
Do you have a working example of using AR QuickLook with custum action button? On my side I can't get it to work really.

**Comment by PixelPartner** [comment]
The custom Action Button/Banner has a single function:It quits AR Quick Look but also emits an event message to the <a> tag it was launched from.Since modelviewer most often creates a blob and neither use a premade .usdz nor has an <a> tag to launch from, you'll find it hard to register a callback function to receive this event anyway.Am 27.02.2024 um 17:30 schrieb Sov3rain ***@***.***>:﻿
Do you have a working example of using AR QuickLook with custum action button? On my side I can't get it to work really.

**Reply by Sov3rain** [reply]
Yeah, I know all of that. I need to be frank here, I'm not using model-viewer to launch AR, I'm using a custom component I wrote. But since the folks here are the only one trying to auto generate an USDZ and pass a blob to AR QuickLook I was hoping to have some enlightenment about the issue with not being able to add parameters to the blob's url.

**Comment by milesgreen** [comment]
@Sov3rain

I'd say this does indeed look like an iOS regression.  

I made a more versatile test suite here:
https://modelviewer-extended-params.glitch.me/

You can turn on / off AutoUSDZ generation, various Quick Look params, and Scene Viewer params. See the ModelViewer src parameters it's generating below. 

Of particular note, I observe on **iOS 17.4.1** that when **AutoUSDZ** is **ON**, and **Apple Pay** is **ON**, clicking through to view ARQL, **no** Apply Pay button is rendered. 

But, switching **AutoUSDZ** to **OFF**, and **Apple Pay** Button to **ON**, clicking through to view in ARQL **does** render the Apply Pay button. So when there is a 'hard' USDZ file, rather than a blob, it's still working. 

Also note that when **AutoUSDZ** is set to **ON**, and **Custom Banner** is set to **ON**, clicking through to view in ARQL **does** indeed show the custom banner. So you can still get custom banner interaction when using blob. 

I ran these tests on several iOS devices:
- iOS 17.4.1 - FAIL: no Apply Pay banner displayed
- iOS 17.2.1 - FAIL: no Apply Pay banner displayed
- iOS 16.6.1 - PASS: Apple Pay banner is displayed
- iOS 15.8.1 - PASS: Apple Pay banner is displayed

So it does indeed look like an iOS regression (rather than ModelViewer). 

If anyone has any other iOS version and could test, it would be interesting to know what version this became an issue.

**Reply by rawwerks** [reply]
thanks for this detective work!

i think apple clearly wants us to all fork out for the headset...all of the AR features are just going to start breaking left and right until we strap that stupid thing to our faces ; )

**Reply by hybridherbst** [reply]
@milesgreen as of iOS 17.5 I can't seem to get any of the combinations to show any kind of overlay on iOS. Is it the same for you?

**Reply by milesgreen** [reply]
@hybridherbst - With iOS 17.5 I was seeing:

USDZ file URL + Apple Banner = WORKS
USDZ file URL + Custom Banner = WORKS
USDZ Blob + Apple Banner = FAILS
USDZ Blob + custom Banner = WORKS

Same for iOS 18 Beta

**Comment by PixelPartner** [comment]
Dear Miles,

Would you mind to repeat this as a bug/regression report at https://bugs.webkit.org/ ?

Great work, BTW !!

Thomas Kumlehn
***@***.***

**Reply by milesgreen** [reply]
Done:
https://bugs.webkit.org/show_bug.cgi?id=271701

**Comment by PixelPartner** [comment]
Thank you
